CRM/Dedupe/Merger.php

```diff
@@ -1631,41 +1632,57 @@ public static function getRowsElementsAndInfo(int $mainID, int $otherID, bool $c
     }
 
     // handle custom fields
-    $mainTree = self::getTree($main['contact_type'], $mainID,
-      $main['contact_sub_type'] ?? NULL,
-      $checkPermissions ? CRM_Core_Permission::EDIT : FALSE
-    );
-    $otherTree = self::getTree($main['contact_type'], $otherID,
-      $other['contact_sub_type'] ?? NULL,
-      $checkPermissions ? CRM_Core_Permission::EDIT : FALSE
-    );
-
-    foreach ($otherTree as $gid => $group) {
-      if (!isset($group['fields'])) {
-        continue;
+    $filters = [
+      'extends' => $main['contact_type'],
+      'is_active' => TRUE,
+      'extends_entity_column_value' => NULL,
+    ];
+    if (!empty($other['contact_sub_type'])) {
+      $filters = array_merge([NULL], $other['contact_sub_type']);
+    }
+    $otherTree = CRM_Core_BAO_CustomGroup::getAll($filters, $checkPermissions ? CRM_Core_Permission::EDIT : NULL);
+
+    foreach ($otherTree as $group) {
+      if (!$group['is_multiple']) {
+        $mainCustomValues = Contact::get(FALSE)
+          ->addWhere('id', '=', $mainID)
+          ->addSelect($group['name'] . '.*')
+          ->execute();
+        $otherCustomValues = Contact::get(FALSE)
+          ->addWhere('id', '=', $otherID)
+          ->addSelect($group['name'] . '.*')
+          ->execute();
       }
-
-      foreach ($group['fields'] as $fid => $field) {
-        $mainContactValue = $mainTree[$gid]['fields'][$fid]['customValue'] ?? NULL;
-        $otherContactValue = $otherTree[$gid]['fields'][$fid]['customValue'] ?? NULL;
+      else {
+        $mainCustomValues = CustomValue::get($group['name'], FALSE)
+          ->addWhere('entity_id', '=', $mainID)
+          ->execute();
+        $otherCustomValues = CustomValue::get($group['name'], FALSE)
+          ->addWhere('entity_id', '=', $otherID)
+          ->execute();
+      }
+      foreach ($group['fields'] as $field) {
+        $fid = $field['id'];
+        $apiFieldName = $group['is_multiple'] ? $group['name'] : "{$group['name']}.{$field['name']}";
         if (in_array($fid, $compareFields['custom'])) {
```

Comment:
So on digging this if - ie the field being in `compareFields` is always FALSE for `is_multiple`

`compareFields['custom']` comes from this code

![image](https://github.com/civicrm/civicrm-core/assets/336308/1470dafa-7ff8-43d3-88fc-baeafddd70bb)


which 
a) never returns any of the multiple custom group fields and
b) would be more sanely replaced by doing a value comparison in this chunk of code that you have added because it's just checking if either contact has a value that is not one of

``` [NULL, '', []];```